packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -0,0 +1,223 @@
+  Point _dragPosition;
+  void _handleDragStart(Point position) {
+    _dragPosition = position;
+  }
+
+  void _handleDragUpdate(double delta) {
+    _dragPosition += new Offset(delta, 0.0);
+    TextPosition position = config.renderObject.getPositionForPoint(_dragPosition);
+
+    if (config.selection.isCollapsed) {
+      config.onSelectionHandleChanged(new TextSelection.fromPosition(position));
+      return;
+    }
+
+    TextSelection newSelection;
+    switch (config.position) {
+      case _TextSelectionHandlePosition.start:
+        newSelection = new TextSelection(
+          baseOffset: position.offset,
+          extentOffset: config.selection.extentOffset
+        );
+        break;
+      case _TextSelectionHandlePosition.end:
+        newSelection = new TextSelection(
+          baseOffset: config.selection.baseOffset,
+          extentOffset: position.offset
+        );
+        break;
+    }
+
+    if (newSelection.baseOffset >= newSelection.extentOffset)
+      return; // don't allow order swapping.
+
+    config.onSelectionHandleChanged(newSelection);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    List<TextSelectionPoint> endpoints = config.renderObject.getEndpointsForSelection(config.selection);
+    Point point;
+    TextSelectionHandleType type;
+
+    switch (config.position) {
+      case _TextSelectionHandlePosition.start:
+        point = endpoints[0].point;
+        type = _chooseType(endpoints[0], TextSelectionHandleType.left, TextSelectionHandleType.right);
+        break;
+      case _TextSelectionHandlePosition.end:
+        // [endpoints] will only contain 1 point for collapsed selections, in
+        // which case we shouldn't be building the [end] handle.
+        assert(endpoints.length == 2);
+        point = endpoints[1].point;
+        type = _chooseType(endpoints[1], TextSelectionHandleType.right, TextSelectionHandleType.left);
+        break;
+    }
+
+    return new GestureDetector(
+      onHorizontalDragStart: _handleDragStart,
+      onHorizontalDragUpdate: _handleDragUpdate,
+      child: new Stack(
+        children: <Widget>[
+          new Positioned(
+            left: point.x,
+            top: point.y,
+            child: _buildChild(context, type)
+          )
+        ]
+      )
+    );
+  }
+
+  TextSelectionHandleType _chooseType(
+    TextSelectionPoint endpoint,
+    TextSelectionHandleType ltrType,
+    TextSelectionHandleType rtlType
+  ) {
+    switch (endpoint.direction) {
+      case TextDirection.ltr:
+        return ltrType;
+      case TextDirection.rtl:
+        return rtlType;
+    }
+  }
+
+  Widget _buildChild(BuildContext context, TextSelectionHandleType type) {
+    Widget child = config.builder(context, type);
+    switch (type) {
+      case TextSelectionHandleType.left:
+        // Shift the child left by 100% of its width, so its top-right corner
+        // touches the selection endpoint.
```

Comment:
This is true for material-design selection handles but isn't true for iOS selection handles.  On iOS, the selection handles are centered (horizontally) at the point of selection.  Why not push this work into the `config.builder` function?